packages/core/src/state-management/state-cache.ts

```diff
@@ -0,0 +1,75 @@
+import { LRUMap } from 'lru_map';
+
+export class StateCache<T> implements Iterable<[string, T]> {
+  readonly volatile: LRUMap<string, T>;
+  readonly durable: Map<string, T>;
+
+  constructor(limit: number, private readonly onEvicted?: (item: T) => void) {
+    this.volatile = new LRUMap(limit);
+    this.volatile.shift = function () {
+      const entry = LRUMap.prototype.shift.call(this);
+      onEvicted?.(entry[1]);
+      return entry;
+    };
+    this.durable = new Map();
+  }
+
+  get(key: string): T | undefined {
+    const fromPersistent = this.durable.get(key);
+    if (fromPersistent) {
+      return fromPersistent;
+    } else {
+      return this.volatile.get(key);
+    }
+  }
+
+  set(key: string, value: T): void {
+    if (this.durable.has(key)) {
+      this.durable.set(key, value);
+    }
+    this.volatile.set(key, value);
```

Comment:
Right now the behavior of `endure` and `set` are different with respect to whether or not a doc should stay in `volatile` if it's also in `durable` (`endure` deletes from `volatile` when adding to `durable`, but `set` will add it back to `volatile` even if it's already in `durable`).

I could see a case for either behavior.  If we keep docs only in one place, so that documents in `durable` do not show up in `volatile`, then that would mean that if an app subscribed to a large number of documents, they'd wind up with a number of documents in-memory in the StateCache  that is equal to the number of subscribed docs PLUS the configured cache size.  I feel like it probably makes sense for docs you've actively subscribed to to count towards your "budget" of docs to keep in the cache.  So in that case I think it probably makes sense to always have docs that are in `durable` also exist in `volatile` as well.  This would also have the nice behavior that unsubscribing from a doc (and thus calling free() on it here in the StateCache) doesn't put the doc to the back of the LRU queue but would allow it to actually be immediately deleted if it hadn't had any updates in a long while.

So that's the way I'd probably go, though that said I could see an argument for going the other way as well.  In either case, we should make the behavior of `endure` and `set` consistent.

CC @oed in case you have an opinion here.